### PR TITLE
[RFC] Eliminate possible race between ALTER_METADATA and MERGE_PARTS

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Eliminate possible race between ALTER_METADATA and MERGE_PARTS (that leads to checksum mismatch - CHECKSUM_DOESNT_MATCH)

Cc: @tavplubix 

Refs: https://github.com/ClickHouse/ClickHouse/pull/28432

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 3d05c4b0a11fe377676101b233e08a8fb2652425 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57755/3d05c4b0a11fe377676101b233e08a8fb2652425/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7170660220">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57755/3d05c4b0a11fe377676101b233e08a8fb2652425/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Storages/MergeTree/MergeFromLogEntryTask.cpp [hunk 1]
```diff
@@ -109,6 +111,18 @@ ReplicatedMergeMutateTaskBase::PrepareResult MergeFromLogEntryTask::prepare()
             };
         }
 
+        int32_t part_metadata_version = source_part_or_covering->getMetadataVersion();
+        if (part_metadata_version > metadata_version)
+        {
+            LOG_DEBUG(log, "Source part metadata version {} is newer then the table metadata version {}. ALTER_METADATA is still in progress.",
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeFromLogEntryTask.cpp:117]: Let's move this below just to make sure that it's actually the source part (and not a covering part)

**Review by tavplubix — COMMENTED** [review]
In general LGTM, but the new test is flaky and the failure of `01079_parallel_alter_detach_table_zookeeper` looks suspicious

File: src/Storages/MergeTree/MergeFromLogEntryTask.cpp (hunk 1 above)
Comment (azat) [reply src/Storages/MergeTree/MergeFromLogEntryTask.cpp]: Done.

**Comment by azat** [comment]
>In general LGTM, but the new test is flaky

Fixed.

> and the failure of 01079_parallel_alter_detach_table_zookeeper looks suspicious

It did not failed this time...

>Stateless tests (release, s3 storage) [1/2] — fail: 1, passed: 2944, skipped: 31

- `02483_test_reverse_dns_resolution`

**Review by tavplubix — APPROVED** [review]

**Pull request merged**